active-rfcs/0000-recursive-components.md

```diff
@@ -0,0 +1,150 @@
+
+Even when the entire tree structure is KNOWN, for anything more than a few levels deep,
+using a recursive Component/function for rendering is the most efficient and elegant approach.
+
+When the tree structure is UNKNOWN, and you want to support N levels deep, using a recursive
+Component/function is the ONLY approach that will work. This is often the case
+when fetching data from an API.
+
+Svelte, which also uses the same "single component per file" pattern has already
+encountered and solved for this issue by exposing the [svelte:self](https://svelte.dev/docs#svelte_self)
+attribute as part of their API.
+
+# Example
+
+We have a set of items we want to render, stored in an object, indexed by id.
+Each item optionally defines an array of childIds.
+
+```
+const items = {
+    "1": {
+        "id": "1",
+        "childIds": ["1-1","1-2"]
+    },
+    "1-1": {
+        id: "1-1",
+        childIds: ["1-1-1", "1-1-2"]
+    },
+    "1-2": {
+        id: "1-2",
+        childIds: []
+    },
+    "1-1-1": {
+        id: "1-1-1",
+        childIds: []
+    },
+    "1-1-2": {
+        id: "1-1-2",
+        childIds: []
+    },
+    "1-1-2-1": {
+        id: "1-1-2-1",
+        childIds: []
+    },
+    "1-1-2-2": {
+        id: "1-1-2-2",
+        childIds: []
+    },
+};
+```
+
+We want to use a recursive component to render the entire tree,
+(or one of its branches) by passing in both the entire tree (the items object)
+and an itemId like this:
+
+```
+// src/components/RecursiveItem.astro
+---
+const {itemId, items} = Astro.props;
+const {childIds} = items[itemId];
+const Self = Astro.self;
+---
+<li>
+    Item: {itemId}
+    {childIds.length ? (
+        <ol>
+            {childIds.map(childId => (
+                <Astro.self itemId={childId} items={items} />
+            ))}
+        </ol>
+    ) : ""}
+</>
+```
+
+Note that in the above example, the `<Self />` component provides
+access to this component's render function, allowing it to
+reference itself.
+
+# Motivation
+
+Handling this use case lets .astro components do things that other component
+frameworks can do, making it a 1st class component framework in its own right.
+
+# Detailed design
+
+The `Astro.self` property exposes the render function of the component.
+Defining a constant in the frontmatter and setting it equal to this allows
+you to name this component whatever you wish.
+
+```
+// src/components/RecursiveItem.astro
+---
+const {itemId, items} = Astro.props;
+const {childIds} = items[itemId];
+const MyRecursiveItem = Astro.self;
+---
+<li>
+    Item: {itemId}
+    {childIds.length ? (
+        <ol>
+            {childIds.map(childId => (
+                <MyRecursiveItem itemId={childId} items={items} />
+            ))}
+        </ol>
+    ) : ""}
+</>
+```
+
+# Alternatives
+
+The other design pattern considered was to mimic the Svelte syntax, using
```

Comment:
Referring to Svelte's solution here helps the proposal, good job! I would be curious to hear what solutions are in other similar frameworks, if any. For example does Vue try to solve this problem, and if so how?